File: sections/64_multiplicative_number_theory/problems/b_summatory_totient/solution.py
import sys

MOD = 1_000_000_007
LIMIT = 1_000_000


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return

    phi = list(range(LIMIT + 1))
    for prime in range(2, LIMIT + 1):
        if phi[prime] == prime:
            for multiple in range(prime, LIMIT + 1, prime):
                phi[multiple] -= phi[multiple] // prime
    for value in range(1, LIMIT + 1):
        phi[value] = (phi[value] + phi[value - 1]) % MOD

    memo: dict[int, int] = {}

    def summatory(n: int) -> int:
        if n <= LIMIT:
            return phi[n]
        if n in memo:
            return memo[n]
        answer = n * (n + 1) // 2 % MOD
        left = 2
        while left <= n:
            quotient = n // left
            right = n // quotient
            answer -= (right - left + 1) * summatory(quotient)
            answer %= MOD
            left = right + 1
        memo[n] = answer
        return answer

    print("\n".join(str(summatory(n)) for n in data[1 : 1 + data[0]]))


if __name__ == "__main__":
    main()

File: sections/64_multiplicative_number_theory/problems/b_summatory_totient/solution.cpp
#include <bits/stdc++.h>
using namespace std;

using int64 = long long;
constexpr int LIMIT = 1'000'000;
constexpr int64 MOD = 1'000'000'007;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<int64> phi(LIMIT + 1);
    iota(phi.begin(), phi.end(), 0);
    for (int p = 2; p <= LIMIT; ++p)
        if (phi[p] == p)
            for (int multiple = p; multiple <= LIMIT; multiple += p)
                phi[multiple] -= phi[multiple] / p;
    for (int i = 1; i <= LIMIT; ++i)
        phi[i] = (phi[i] + phi[i - 1]) % MOD;

    unordered_map<int64, int64> memo;
    function<int64(int64)> summatory = [&](int64 n) -> int64 {
        if (n <= LIMIT) return phi[n];
        if (auto it = memo.find(n); it != memo.end()) return it->second;
        int64 answer = n % MOD * ((n + 1) % MOD) % MOD;
        answer = answer * ((MOD + 1) / 2) % MOD;
        for (int64 left = 2, right; left <= n; left = right + 1) {
            int64 quotient = n / left;
            right = n / quotient;
            int64 width = (right - left + 1) % MOD;
            answer = (answer - width * summatory(quotient)) % MOD;
        }
        if (answer < 0) answer += MOD;
        return memo[n] = answer;
    };

    int q;
    cin >> q;
    while (q--) {
        int64 n;
        cin >> n;
        cout << summatory(n) << '\n';
    }
}
